transport/x/grpc/handler.go

```diff
@@ -0,0 +1,185 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package grpc
+
+import (
+	"bytes"
+	"fmt"
+	"time"
+
+	"go.uber.org/multierr"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/encoding/x/protobuf"
+	"go.uber.org/yarpc/internal/errors"
+	"go.uber.org/yarpc/internal/request"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+type handler struct {
+	procedureServiceName string
+	serviceName          string
+	methodName           string
+	router               transport.Router
+	onewayErrorHandler   func(error)
+}
+
+func newHandler(
+	procedureServiceName string,
+	serviceName string,
+	methodName string,
+	router transport.Router,
+	onewayErrorHandler func(error),
+) *handler {
+	return &handler{
+		procedureServiceName,
+		serviceName,
+		methodName,
+		router,
+		onewayErrorHandler,
+	}
+}
+
+func (h *handler) handle(
+	server interface{},
+	ctx context.Context,
+	decodeFunc func(interface{}) error,
+	interceptor grpc.UnaryServerInterceptor,
+) (interface{}, error) {
+	transportRequest, err := h.getTransportRequest(ctx, decodeFunc)
+	if err != nil {
+		return nil, err
+	}
+	if interceptor != nil {
+		return interceptor(
+			ctx,
+			transportRequest,
+			&grpc.UnaryServerInfo{
+				noopGrpcStruct{},
+				h.getFullMethod(),
+			},
+			func(ctx context.Context, request interface{}) (interface{}, error) {
+				transportRequest, ok := request.(*transport.Request)
+				if !ok {
+					return nil, fmt.Errorf("expected *transport.Request, got %T", request)
+				}
+				return h.call(ctx, transportRequest)
+			},
+		)
+	}
+	return h.call(ctx, transportRequest)
+}
+
+func (h *handler) getTransportRequest(ctx context.Context, decodeFunc func(interface{}) error) (*transport.Request, error) {
+	md, ok := metadata.FromContext(ctx)
+	if md == nil || !ok {
+		return nil, fmt.Errorf("cannot get metadata from ctx: %v", ctx)
+	}
+	caller, err := getCaller(md)
+	if err != nil {
+		return nil, err
+	}
+	if caller == "" {
+		caller = h.serviceName
+	}
+	encoding, err := getEncoding(md)
+	if err != nil {
+		return nil, err
+	}
+	if encoding == "" {
+		encoding = protobuf.Encoding
+	}
+	service, err := getService(md)
```

Comment:
I removed all these functions in favor of `transport.Request` functions - this is something I wanted to do anyways